src/garage/tf/algos/batch_polopt.py

```diff
@@ -232,40 +236,41 @@ def process_samples(self, itr, paths):
         valids = [np.ones_like(path['returns']) for path in paths]
         valids = tensor_utils.pad_tensor_n(valids, max_path_length)
 
-        average_discounted_return = (np.mean(
-            [path['returns'][0] for path in paths]))
-
-        undiscounted_returns = [sum(path['rewards']) for path in paths]
-        self.episode_reward_mean.extend(undiscounted_returns)
+        lengths = np.asarray([v.sum() for v in valids])
 
         ent = np.sum(self.policy.distribution.entropy(agent_infos) *
                      valids) / np.sum(valids)
 
+        undiscounted_returns = self.evaluate_performance(
+            itr,
+            dict(env_spec=self.env_spec,
+                 observations=obs,
+                 actions=actions,
+                 rewards=rewards,
+                 terminals=terminals,
+                 env_infos=env_infos,
+                 agent_infos=agent_infos,
+                 lengths=lengths,
+                 discount=self.discount,
+                 episode_reward_mean=self.episode_reward_mean))
+
+        tabular.record('Entropy', ent)
+        tabular.record('Perplexity', np.exp(ent))
+
         samples_data = dict(
             observations=obs,
             actions=actions,
             rewards=rewards,
             baselines=baselines,
             returns=returns,
             valids=valids,
+            lengths=lengths,
             agent_infos=agent_infos,
             env_infos=env_infos,
             paths=paths,
             average_return=np.mean(undiscounted_returns),
         )
 
-        tabular.record('Iteration', itr)
-        tabular.record('AverageDiscountedReturn', average_discounted_return)
-        tabular.record('AverageReturn', np.mean(undiscounted_returns))
-        tabular.record('Extras/EpisodeRewardMean',
-                       np.mean(self.episode_reward_mean))
-        tabular.record('NumTrajs', len(paths))
-        tabular.record('Entropy', ent)
-        tabular.record('Perplexity', np.exp(ent))
```

Comment:
it's fine to still log `Entropy` and `Perplexity` here -- did you move these somewhere else?

Reply:
It's right above (L257).